src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioActiveDocumentTracker.cs

```diff
@@ -0,0 +1,290 @@
+        /// Collection of all asynchronous tasks that are started by this service. This should only be tasks that are implicitly
+        /// async as we fetch services from the <see cref="IAsyncServiceProvider"/>, and are used to ensure we wait for them during shutdown.
+        /// These should not be waited for in calls to <see cref="GetActiveDocument(Workspace)"/> or <see cref="GetVisibleDocuments(Workspace)"/>
+        /// because those are expected to not be jumping to the UI thread per traditional Roslyn semantics and may deadlock.
+        /// </summary>
+        private readonly JoinableTaskCollection _asyncTasks;
+
+        private ImmutableList<FrameListener> _visibleFrames = ImmutableList<FrameListener>.Empty;
+        private IVsWindowFrame _activeFrame;
+
+        [ImportingConstructor]
+        public VisualStudioActiveDocumentTracker(IThreadingContext threadingContext, [Import(typeof(SVsServiceProvider))] IAsyncServiceProvider asyncServiceProvider, IVsEditorAdaptersFactoryService editorAdaptersFactoryService)
+            : base(threadingContext, assertIsForeground: false)
+        {
+            _threadingContext = threadingContext;
+            _editorAdaptersFactoryService = editorAdaptersFactoryService;
+            _asyncTasks = new JoinableTaskCollection(threadingContext.JoinableTaskContext);
+            _asyncTasks.Add(threadingContext.JoinableTaskFactory.RunAsync(async () =>
+            {
+                await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+                var monitorSelectionService = (IVsMonitorSelection)await asyncServiceProvider.GetServiceAsync(typeof(SVsShellMonitorSelection));
+
+                if (ErrorHandler.Succeeded(monitorSelectionService.GetCurrentElementValue((uint)VSConstants.VSSELELEMID.SEID_DocumentFrame, out var value)))
+                {
+                    if (value is IVsWindowFrame windowFrame)
+                    {
+                        TrackWindowFrame(windowFrame);
+                        _activeFrame = windowFrame;
+                    }
+                }
+
+                monitorSelectionService.AdviseSelectionEvents(this, out var _);
+            }));
+        }
+
+        public void Dispose()
+        {
+            _asyncTasks.Join();
+        }
+        
+        /// <summary>
+        /// Raised when the set of window frames being tracked changes, which means the results from <see cref="GetActiveDocument"/> or <see cref="GetVisibleDocuments"/> may change.
+        /// </summary>
+        public event EventHandler DocumentsChanged;
+
+        /// <summary>
+        /// Returns the <see cref="DocumentId"/> of the active document in a given <see cref="Workspace"/>.
+        /// </summary>
+        public DocumentId GetActiveDocument(Workspace workspace)
+        {
+            ThisCanBeCalledOnAnyThread();
+
+            var visibleFramesSnapshot = _visibleFrames;
+            if (_activeFrame == null || visibleFramesSnapshot.IsEmpty)
```

Comment:
since this can be called on any thread, should we not make a local copy of _activeFrame as well?

Reply:
I've done a local clone and added a comment. The decision was it doesn't matter: this API is unsynchronized relative to the UI thread. If _activeFrame is mutated in the middle from one frame to another, it means it might not say either is active. That's not really a problem I don't think -- the event will get fired and everything will be fine later.